### PR TITLE
add assertions tests

### PR DESCRIPTION
**Pull request by Robert-Costello**
## Technical Summary

relates to https://github.com/dimagi/commcare-core/pull/1267

## Safety Assurance
just tests

### Automated test coverage

is only tests

### QA Plan

relates to this QA ticket: https://dimagi-dev.atlassian.net/browse/QA-5258
### Special deploy instructions

- [x] This PR can be deployed after merge with no further considerations.

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations.

### Review

- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change.

### CODE REVIEW

File: src/test/java/org/commcare/formplayer/tests/MenuAssertionsTest.java [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+    }
+
+
+    private <T extends BaseResponseBean> Response<T> navigate(
+            String[] selections, Class<T> responseClass) {
+        String installReference = Installer.getInstallReference(ASSERTIONS_APP);
+        SessionNavigationRequest<T> request = new SessionNavigationRequest<>(
+                mockMvc, responseClass, installReference);
+        SessionNavigationBean bean = request.getNavigationBean(selections);
+        return request.requestWithBean(bean);
```
Comment (snopoke) [inline src/test/java/org/commcare/formplayer/tests/MenuAssertionsTest.java:87]: nit: you can simplify this a bit:
```suggestion
        return request.request(selections);
```

**Review by snopoke — APPROVED** [review]

File: src/test/java/org/commcare/formplayer/tests/MenuAssertionsTest.java [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+
+
+    @Test
+    public void testMenuAssertions() {
+        // Check passing assertion
+        Response<CommandListResponseBean> response0 = navigate(new String[]{"0"},
+                CommandListResponseBean.class);
+        assertNotNull(response0);
+        // Check failing assertion
+        assertThrows(Exception.class, () -> {
```
Comment (shubham1g5) [inline src/test/java/org/commcare/formplayer/tests/MenuAssertionsTest.java:74]: can we also verify that exception has the right message as per assertions here.

**Review by shubham1g5 — CHANGES_REQUESTED** [review]
Hey Robert, the PR looks good. Wondering if it's possible to comment on the PR the changes in `.ccz` file for the sake of visibility. 

Also we should add tests for a few different type of edge cases we discovered during testing this - 
1. Assertions that use external instances `instance(x)` 
2. Assertions on root menu with instance connectors defined in `menu` 

It would also be great to have both positive and negative test cases around assertions. For eg we can write tests testing `count(instance('commcaresession')/session/user/data/project_role) > 0` for 2 different users (aka with 2 different restores) one with `project_role` and other without and check assertions accordingly.

**Comment by Robert-Costello** [comment]
> Wondering if it's possible to comment on the PR the changes in `.ccz` file for the sake of visibility.

in the suite file I added assertions and instance references to the groups fixture already being used. 

menu 0:
\<menu id="m0">
    &nbsp;&nbsp;&nbsp;&nbsp;\<text>
     &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; \<locale id="modules.m0"/>
   &nbsp;&nbsp;&nbsp;&nbsp; \</text>
    &nbsp;&nbsp;&nbsp;&nbsp;\<instance id="groups" src="jr://fixture/user-groups"/>
    &nbsp;&nbsp;&nbsp;&nbsp;\<assertions>
     &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; \<assert test="count(instance('groups')/groups/group) = 1">
        &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;\<text>
        &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp;&nbsp;\<locale id="case_sharing.exactly_one_group"/>
       &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; \</text>
     &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; \</assert>
    &nbsp;&nbsp;&nbsp;&nbsp;\</assertions>
    &nbsp;&nbsp;&nbsp;&nbsp;\<command id="m0-f0"/>...
  \</menu>

menu 1:
\<menu id="m1">
   &nbsp;&nbsp;&nbsp;&nbsp;\<text>
     &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;\<locale id="modules.m1"/>
    &nbsp;&nbsp;&nbsp;&nbsp;\</text>
    &nbsp;&nbsp;&nbsp;&nbsp;\<instance id="groups" src="jr://fixture/user-groups"/>
    &nbsp;&nbsp;&nbsp;&nbsp;\<assertions>
     &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;\<assert test="count(instance('groups')/groups/group) = 2">
     &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp;&nbsp; \<text>
         &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;\<locale id="case_sharing.exactly_one_group"/>
       &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; \</text>
     &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; \</assert>
   &nbsp;&nbsp;&nbsp;&nbsp; \</assertions>
  &nbsp;&nbsp;&nbsp;&nbsp; \<command id="m1-f0"/>...
\</menu>

File: src/test/java/org/commcare/formplayer/tests/MenuAssertionsTest.java (hunk 2 above)
Comment (Robert-Costello) [reply src/test/java/org/commcare/formplayer/tests/MenuAssertionsTest.java]: verified here https://github.com/dimagi/formplayer/pull/1411/commits/19a982c29550c2e8bbc520bca803cbf84eb764a2

File: src/test/java/org/commcare/formplayer/tests/MenuAssertionsTest.java (hunk 1 above)
Comment (Robert-Costello) [reply src/test/java/org/commcare/formplayer/tests/MenuAssertionsTest.java]: simplified here https://github.com/dimagi/formplayer/pull/1411/commits/cd051b4cf20b6e5e09842c3a64c29c332c348c2d

**Comment by Robert-Costello** [comment]
These tests are passing locally. Not sure why they are failing here. 
`at org.commcare.formplayer.tests.CasePaginationTests.checkDetailTemplates(CasePaginationTests.java:94)`
 `at org.commcare.formplayer.tests.CasePaginationTests.testPagination(CasePaginationTests.java:65)`
@shubham1g5 any ideas?

**Comment by shubham1g5** [comment]
They seem to fail locally as well for me. You probably wanna cross-check changes in `basic.ccz`.

**Comment by Robert-Costello** [comment]
It turns out that it's the submodule ref that is causing the failures. When I switch to the formplayer branch locally, I get the same failures.

File: src/test/resources/restores/project_role_admin.xml [hunk 3]
```diff
@@ -0,0 +1,4496 @@
+<OpenRosaResponse xmlns="http://openrosa.org/http/response">
+<message nature="ota_restore_success">Successfully restored account vl1!</message>
+<Sync xmlns="http://commcarehq.org/sync">
+    <restore_id>9f7c4e0a3a8ca4e2ecf54444fb7e70cf</restore_id>
+</Sync>
+<Registration xmlns="http://openrosa.org/user/registration">
```
Comment (shubham1g5) [inline src/test/resources/restores/project_role_admin.xml:6]: Can we trim the file to only contain what's necessary for the test. I believe you should be able to remove most of the info in this file for the assertion test.

Comment (Robert-Costello) [reply src/test/resources/restores/project_role_admin.xml:6]: restore updated here https://github.com/dimagi/formplayer/pull/1411/commits/0663863504d0e7c8eae118273b59572bcf3dc3c5

**Review by shubham1g5 — APPROVED** [review]
💯
